src/java/org/apache/cassandra/db/compaction/unified/AdaptiveController.java

```diff
@@ -0,0 +1,330 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.db.compaction.unified;
+
+import java.util.Arrays;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+
+import javax.annotation.Nullable;
+
+import com.google.common.annotations.VisibleForTesting;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.cassandra.db.compaction.UnifiedCompactionStrategy;
+import org.apache.cassandra.exceptions.ConfigurationException;
+import org.apache.cassandra.utils.FBUtilities;
+import org.apache.cassandra.utils.SystemTimeSource;
+import org.apache.cassandra.utils.TimeSource;
+
+/**
+ * The adaptive compaction controller dynamically calculates the optimal scaling factor W.
+ * <p/>
+ * Generally it tries to find a local minimum for the total IO cost that is projected
+ * by the strategy. The projected IO cost is composed by two parts: the read amplification,
+ * which is weighted by the number of partitions read by the user, and the write amplification, which
+ * is weighted by the number of bytes inserted into memtables. Other parameters are also considered, such
+ * as the cache miss rate and the time it takes to read and write from disk. See also the comments in
+ * {@link CostsCalculator}.
+ *
+ * Design doc: TODO: link to design doc or SEP
+ */
+public class AdaptiveController extends Controller
+{
+    private final static Logger logger = LoggerFactory.getLogger(AdaptiveController.class);
+
+    /** The starting value for W */
+    final static String STARTING_W = "adaptive_starting_w";
+    private final static int DEFAULT_STARTING_W = Integer.getInteger(PREFIX + STARTING_W, 0);
+
+    /** The minimum valid value for W */
+    final static String MIN_W = "adaptive_min_w";
+    private final static int DEFAULT_MIN_W = Integer.getInteger(PREFIX + MIN_W, -10);
+
+    /** The maximum valid value for W */
+    final static String MAX_W = "adaptive_max_w";
+    private final static int DEFAULT_MAX_W = Integer.getInteger(PREFIX + MIN_W, 36);
+
+    /** The interval for periodically checking the optimal value for W */
+    final static String INTERVAL_SEC = "adaptive_interval_sec";
+    private final static int DEFAULT_INTERVAL_SEC = Integer.getInteger(PREFIX + INTERVAL_SEC, 300);
+
+    /** The gain is a number between 0 and 1 used to determine if a new choice of W is better than the current one */
+    final static String GAIN = "adaptive_gain";
```

Comment:
"Gain" is associated with some kind of amplification. "Threshold" is a better term here.